auth/provide.go

```diff
@@ -0,0 +1,79 @@
+package auth
+
+import (
+	"reflect"
+
+	"github.com/go-kit/kit/log"
+
+	"github.com/justinas/alice"
+	"github.com/xmidt-org/bascule/basculehttp"
+	"go.uber.org/fx"
+)
+
+type primaryChainIn struct {
+	fx.In
+	SetLogger   alice.Constructor `name:"primary_alice_set_logger" optional:"true"`
+	Constructor alice.Constructor `name:"primary_alice_constructor" optional:"true"`
+	Enforcer    alice.Constructor `name:"primary_alice_enforcer" optional:"true"`
+	Listener    alice.Constructor `name:"primary_alice_listener" optional:"true"`
+}
+
+type primaryProfileIn struct {
+	fx.In
+	Logger  log.Logger
+	Profile *profile `name:"primary_profile" optional:"true"`
+}
+
+// ProvidePrimaryServerChain provides the auth alice.Chain for the primary server.
+func ProvidePrimaryServerChain() fx.Option {
+	return fx.Options(
+		LogOptionsProvider{ServerName: "primary"}.Provide(),
+		ProvidePrimaryBasculeConstructor(),
+		ProvidePrimaryBasculeEnforcer(),
+		ProvidePrimaryTokenFactory(),
+		fx.Provide(
+			profileProvider{ServerName: "primary"}.Annotated(),
+			BasculeMetricsListenerProvider{ServerName: "primary"}.Annotated(),
+
+			fx.Annotated{
+				Name: "primary_alice_listener",
+				Target: func(in PrimaryBasculeMetricListenerIn) alice.Constructor {
+					return basculehttp.NewListenerDecorator(in.Listener)
+				},
+			},
+			fx.Annotated{
+				Name: "primary_auth_chain",
+				Target: func(in primaryChainIn) alice.Chain {
+					if anyNil(in.SetLogger, in.Constructor, in.Enforcer, in.Listener) {
```

Comment:
I don't think any of these can be nil (see previous comments in similar checks in constructor.go and enforcer.go), but I'm also not sure that they should be.  The previous auth code always made the bascule middleware or errored out, so as far as I can tell we always required one kind of authentication? So none of these should be optional.

But if we want to allow setting up the server with no auth, then this logic is okay (but we should change other things so it's possible).

Reply:
I added a comment about this in the comment above. Yeah, it's fail open now and I believe our other services are fail closed with the unsupported key error 👍

Reply:
To follow that functionality, none of these should ever be nil/optional.  That's basically how it is in older applications; we always create them, even if they don't allow any requests.  In that case, I don't think we should ever return nil in the constructor and enforcer functions - we always should be creating something.